### PR TITLE
INTLY-8120 Strip down rhmi_status metric to only have the overall stage

### PR DESCRIPTION
**Pull request by david-martin**
# Description
https://issues.redhat.com/browse/INTLY-8120

Removes all labels from the `rhmi_status` metric so the number of series is kept to a minimum (more info in jira & linked document).
These changes will ensure a limit of 8 series (1 for each status).
A metric series will be exposed at all times for the current stage, with a value of 1.
For example:

```
# HELP rhmi_status RHMI status of an installation
# TYPE rhmi_status gauge
rhmi_status{stage="complete"} 1
```

An expression to determine if the RHMI install is complete would be:

```
rhmi_status{stage="complete"} == 1
```

If you want to get the currently active stage:

```
rhmi_status == 1
```
and use the value of the `stage` label

This approach seems more inline with how metrics are used to expose non numerical data, but comes with caveats (hence the need to limit the possible values of labels).

More info on metric cardinality in https://www.robustperception.io/cardinality-is-key

To verify this change with the operator running locally, run:

```
while true; do curl localhost:8383/metrics|grep status & sleep 1; done
``` 
and verify the output looks similar to below, with the current stage having a value of 1 during the installation

```
# HELP rhmi_status RHMI status of an installation
# TYPE rhmi_status gauge
rhmi_status{stage="bootstrap"} 1
```

The stage should change as the CR status.stage field changes

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update 
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

### CODE REVIEW

**Comment by david-martin** [comment]
@jjaferson I'd appreciate your review here as you're familiar with the metrics being modified.

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -87,64 +84,12 @@ func SetRHMIInfo(installation *integreatlyv1alpha1.RHMI) {
...
-		customMetrics.Registry.MustRegister(RHMIStatus)
-	}
-}
-
-func SanitizeForPrometheusLabel(productName integreatlyv1alpha1.ProductName) string {
-	if productName == integreatlyv1alpha1.Product3Scale {
-		productName = "threescale"
+	RHMIStatus.Reset()
+	if string(installation.Status.Stage) != "" {
+		RHMIStatus.With(prometheus.Labels{"stage": string(installation.Status.Stage)}).Set(float64(1))
```
Comment (smarterclayton) [inline pkg/metrics/metrics.go:91]: If you have empty stage, you probably still want to report the metric, right?  Is there meaning to "installed and this code is running but no stage is set" such that not reporting the value makes sense?

Comment (smarterclayton) [reply pkg/metrics/metrics.go:91]: I might suggest avoiding reset and just setting the value always to `"stage": string(installation.Status.Stage)`.  That feels more natural use of the "status" label.  If stage == "" has special meaning, and you want the value of the metric to be zero, then it makes the code more explicit with an else block (instead of reset).

Comment (david-martin) [reply pkg/metrics/metrics.go:91]: @maleck13 or @philbrookes What are your thoughts on this?
What does an empty stage mean?

> I might suggest avoiding reset and just setting the value always to `"stage": string(installation.Status.Stage)`

@smarterclayton I might be using the prom library incorrectly, but if the metric isn't reset it results in the number of series going up by 1 each time the stage progresses. It's accumulative.

e.g.
scrape after startup
```
rhmi_status{stage=""} 1
```
scrape after <1m running
```
rhmi_status{stage=""} 1
rhmi_status{stage="Preflight Checks"} 1
```
scrape after >2m running
```
rhmi_status{stage=""} 1
rhmi_status{stage="Preflight Checks"} 1
rhmi_status{stage="cloud-resources"} 1
```
and so on until complete & there are 8 series.

The series seem to be held in memory.
So if the operator is killed/restarted, the metric is emptied of all series and starts building them up from whatever the current stage is.
This seems like odd/undesired behaviour vs. only have 1 series exposed at all times with the current stage, or have 0's for series unless it's the current stage.

Comment (philbrookes) [reply pkg/metrics/metrics.go:91]: stage being empty is what would happen before bootstrap has started. This stage shouldn't last long, but could be prolonged by failing preflight checks. I'd be reasonably happy with the statement `"" == "preflight"`in this sense. 

@Boomatang you initially implemented the stage field in the CR, do you agree?

Comment (Boomatang) [reply pkg/metrics/metrics.go:91]: The stage field in the CR is set a the start of each reconcile stage if that stage is in progress. When starting there is only a very short time that the stage has no value till the preflight checks start. If for some reason the preflight checks fail the CR stage value will stay as `"Prefight Checks"`. I don't see any way once the CR stage is set that it can go back to been a blank value.

The one caveat to be aware of is the stage value is set in the CR if the reconciler reports a stage is in progress. For example RHMI is full installed and the CR stage is marked as "completed". During the reconcile loop the "cloud-resources" actions start. Firstly the stage checks if the reconciler reports a "in progress" status for the "cloud-resources" and updates the CR if required. If during the rest of the "cloud-resources" actions the status gets set to "in progress" the CR stage will not be updated till the next pass of the reconcile loop. This is also true when doing an install. The CR stage may report the previous stage while starting to install the next stage but will be updated with the correct values on the next pass of the reconcile loop.

In short I agree with @philbrookes the length of time the CR stage value can be empty is so short there is no need to do anything with this. And sense this is a metric been reported my guess is we would have gotten pass the "preflight" stage I don't see how the CR stage value can ever be set back to a blank value.

**Comment by david-martin** [comment]
@jjaferson could i get another lgtm? rebased

**Comment by david-martin** [comment]
flake https://github.com/integr8ly/integreatly-operator/pull/885 https://issues.redhat.com/browse/INTLY-8433

**Comment by david-martin** [comment]
last failure is potentially a new test flake

```
        --- FAIL: TestIntegreatly/integreatly/Test_RHMI_installation_CR_metric (4.47s)
            rhmicr_metrics.go:46: rhmi_status metric stage does not match current stage: solution-explorer != complete
```

I believe this failure is possible if the latest scrape of metrics in prometheus has an out of date status compared to the current status in the RHMI CR.
Will investigate a way to avoid this.

**Comment by david-martin** [comment]
> I believe this failure is possible if the latest scrape of metrics in prometheus has an out of date status compared to the current status in the RHMI CR.

@jjaferson Do you have any ideas here why the metric from the most recent scrape might not match the current value in the RHMI CR?
I thought it might be because of a time lag, but we're calling the metrics endpoint directly from inside the operator pod *after* we've read the RHMI CR.
However, in this failure the CR status was `complete` but the metric was `solution-explorer`

**Comment by david-martin** [comment]
@philbrookes can you give this an lgtm again.
I've applied the discussed changes and they solved the failing test with the incorrect stage.
